### PR TITLE
Authd replacement configurations QA

### PR DESCRIPTION
**Pull request by palaciosjeremias**
|Related issue|
|---|
|#2050|

## Description
As part of the epic https://github.com/wazuh/wazuh-qa/issues/2050, we created all the necessary changes to cope with the new feature developed in https://github.com/wazuh/wazuh/issues/9550.
In this PR we:
- Fixed legacy Wazuh-DB test cases that didn´t expect the new disconnection_time in **gobal.db**.
- Fixed legacy Authd tests that expected a raw replacement of the agent and now were rejected.
- Adapted all the **force_insert** and **force_time** configurations to the new **force** block .
- Reformulated key_hash test to check the correct parse of the **K** option in the enrollment message.
- Created a new test to validate the correct behavior of Authd with all the valid force **options** configuration (test_authd_force_options.py).
- Created a new test to validate the error handling of Authd with invalid force **options** (test_authd_force_options_invalid_config.py).

## Tests

- [x] Proven that tests **pass** when they have to pass.
- [x] Proven that tests **fail** when they have to fail.
- [x] Python codebase satisfies PEP-8 style style guide. `pycodestyle --max-line-length=120 --show-source --show-pep8 file.py`.
- [x] Python codebase is documented following the DocGenerator schema v2.0.
- [x] DocGenerator sanity check is executed in the new code without errors.

### CODE REVIEW

**Comment by palaciosjeremias** [comment]
Successful execution of authd tests: https://ci.wazuh.info/job/Test_integration/15012/
Successful execution of wazuhdb tests: https://ci.wazuh.info/job/Test_integration/15014/

**Comment by MiguelazoDS** [comment]
DocGenerator sanity without errors.

image

File: tests/integration/test_authd/force_options/README.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# force options tests
+These tests aim to check the correct behavior of Authd under the different possible configurations of the force block.
```
Comment (juliamagan) [inline tests/integration/test_authd/force_options/README.md:2]: This file shouldn't be here

**Review by juliamagan — CHANGES_REQUESTED** [review]

**Comment by juliamagan** [comment]
## 05/11/2021

### Package
| Version | Revision | Link|
|---|---|---|
4.3.0|40301|https://packages-dev.wazuh.com/warehouse/pullrequests/4.3/rpm/var/wazuh-manager-4.3.0-0.commitf39f878.x86_64.rpm|


## Testing

### tests/integration/test_authd

|  OS    | Local |Jenkins     | Notes
|---      |---    |---    |---  |
| PS1   | 🔵   | | 
| PS2   |  🔵   | | |
| PS3   |   🔵  |   | |


* * * 

- :green_circle:: All pass
- :yellow_circle:: Some warnings
- :red_circle:: Some errors/fails
- :large_blue_circle:: In progress

File: deps/wazuh_testing/wazuh_testing/authd.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+            return result, err_msg
+
+        message = response[1]
+        if 'message' in expected:
+            if re.match(expected['message'], message) is None:
+                return 'error', f"Invalid 'message': '{message}' received, '{expected['message']}' expected"
+    else:
+        raise Exception('Invalid expected status')
+
+    return result, err_msg
```
Comment (Rebits) [inline deps/wazuh_testing/wazuh_testing/authd.py:132]: This function is a little hard to read and it has code repetition. In future revisions of this module we could use something like the following:

```suggestion
def get_authd_response_values(response):
    authd_response_dict = {}
    try: 
        response = response.split(sep=' ', maxsplit=1)
        status, message = response[0:1]
        
        authd_response_dict['status'] = status
        authd_response_dict['message'] = message
        
        if status == 'success':
            agent_key = message.split('\'')[1::2][0].split()
            id, name, ip, key = agent_key[0:4]
            authd_response_dict['id'] = id
            authd_response_dict['name'] = name
            authd_response_dict['ip'] = ip
            authd_response_dict['key'] = key
    except:
        raise IndexError(f"Authd response {response} does not contains expected values: {str(values)}")
    return authd_response_dict
    
        
def validate_authd_response(response, expected):
    """
    Validates if the different items of an Authd response are as expected. Any item inexistent in expected won't
    be validated.

    Args:
        response (str): The Authd response to be validated.
        expected (dict): Dictionary with the items to validate.
    """
    authd_response_dict = get_authd_response_values(response)

    result = 'success'
    err_msg = ''
    
    for expected_key in expected.keys():
        if expected_key == 'message':
            if re.match(expected[expected_key], authd_response_dict[message]) is None:
                return 'error', f"Invalid 'message': '{message}' received, '{expected['message']}' expected"
        elif expected_key in ['status', 'id', 'name', 'ip', 'key']:
            if expected_key == 'status':
                expected_value = 'OSSEC' if expected['status'] == 'success' else 'ERROR'
            else:
                expected_value = expected[expected_key]
            
            result, err_msg = validate_argument(authd_response_dict[expected_value], expected_value, expected_key)
        else:
            raise Exception('Invalid expected status')

    return result, err_msg
```

File: deps/wazuh_testing/wazuh_testing/authd.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+def clean_agents_from_db():
+    """
+    Clean agents from DB
+    """
+    command = 'global sql DELETE FROM agent WHERE id != 0'
+    try:
+        query_wdb(command)
+    except Exception:
+        raise Exception('Unable to clean agents')
```
Comment (Rebits) [inline deps/wazuh_testing/wazuh_testing/authd.py:143]: Maybe is more intuitive to move this to `wazuh_db` due this is not properly an authd related function.

File: deps/wazuh_testing/wazuh_testing/authd.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+
+def insert_agent_in_db(id=1, name="TestAgent", ip="any", registration_time=0, connection_status=0,
+                       disconnection_time=0):
```
Comment (Rebits) [inline deps/wazuh_testing/wazuh_testing/authd.py:147]:
```suggestion
def insert_agent_in_db(id=1, name='TestAgent', ip='any', registration_time=0, connection_status=0,
                       disconnection_time=0):
```

File: deps/wazuh_testing/wazuh_testing/authd.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    Write agent in global.db
+    """
+    insert_command = f'global insert-agent {{"id":{id},"name":"{name}","ip":"{ip}","date_add":{registration_time}}}'
+    update_command = f'global sql UPDATE agent SET connection_status = "{connection_status}",\
+                       disconnection_time = "{disconnection_time}" WHERE id = {id};'
+    try:
+        query_wdb(insert_command)
+        query_wdb(update_command)
+    except Exception:
+        raise Exception(f'Unable to add agent {id}')
```
Comment (Rebits) [inline deps/wazuh_testing/wazuh_testing/authd.py:158]: Same that before, these functions could be useful for the rest of the modules and they are more related with `wazuh_db` module.

Comment (Rebits) [inline deps/wazuh_testing/wazuh_testing/authd.py:158]:
```suggestion
def insert_agent_in_db(id=1, name="TestAgent", ip="any", registration_time=0, connection_status=0,
                       disconnection_time=0):
    """
    Write agent in global.db
    """
    insert_command = f'global insert-agent {{"id":{id},"name":"{name}","ip":"{ip}","date_add":{registration_time}}}'
    update_command = f'global sql UPDATE agent SET connection_status = "{connection_status}",\
                       disconnection_time = "{disconnection_time}" WHERE id = {id};'
    try:
        query_wdb(insert_command)
        query_wdb(update_command)
    except Exception:
        raise Exception(f"Unable to add agent {id}")
```

File: deps/wazuh_testing/wazuh_testing/authd.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        else:
+            registration_time = time_now
+
+        # Write agent in client.keys
+        keys_file.write(f'{id} {name} {ip} {key}\n')
+
+        # Write agent in global.db
+        insert_agent_in_db(id, name, ip, registration_time, connection_status, disconnection_time)
+
+    keys_file.close()
```
Comment (Rebits) [inline deps/wazuh_testing/wazuh_testing/authd.py:197]: Fixtures should be placed in `test_authd` conftest.

File: deps/wazuh_testing/wazuh_testing/tools/services.py [hunk 7]
```diff
@@ -205,6 +205,7 @@ def check_daemon_status(target_daemon=None, running_condition=True, timeout=10,
         elapsed_time = 0
 
         while elapsed_time < timeout and not condition_met:
+
```
Comment (Rebits) [inline deps/wazuh_testing/wazuh_testing/tools/services.py:208]:
```suggestion
```

File: tests/integration/test_authd/conftest.py [hunk 8]
```diff
@@ -1,9 +1,16 @@
...
 from wazuh_testing.tools.file import truncate_file
 from wazuh_testing.tools.monitoring import FileMonitor, make_callback, AUTHD_DETECTOR_PREFIX
-from wazuh_testing.tools.services import control_service
-from authd import DAEMON_NAME
+from wazuh_testing.tools.configuration import write_wazuh_conf, get_wazuh_conf, set_section_wazuh_conf,\
+                                              load_wazuh_configurations
+from wazuh_testing.tools.services import control_service, check_daemon_status, delete_dbs
+from wazuh_testing.tools.monitoring import QueueMonitor
+
+
```
Comment (Rebits) [inline tests/integration/test_authd/conftest.py:12]:
```suggestion
```

File: tests/integration/test_authd/conftest.py [hunk 9]
```diff
@@ -1,9 +1,16 @@
 import pytest
+import os
+import yaml
 from wazuh_testing.tools import LOG_FILE_PATH, CLIENT_KEYS_PATH
```
Comment (Rebits) [inline tests/integration/test_authd/conftest.py:4]:
```suggestion

from wazuh_testing.tools import LOG_FILE_PATH, CLIENT_KEYS_PATH
```

File: tests/integration/test_authd/force_options/test_authd_force_options.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    - Red Hat 8
+    - Red Hat 7
+    - Red Hat 6
+
+tags:
+    - enrollment
+'''
+import os
+import time
+import pytest
```
Comment (Rebits) [inline tests/integration/test_authd/force_options/test_authd_force_options.py:52]:
```suggestion
import pytest

```

File: tests/integration/test_authd/force_options/test_authd_force_options.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+# Configurations
+configurations = load_wazuh_configurations(configurations_path, __name__)
+local_internal_options = {'authd.debug': '2'}
+
+# Tests
+tests = []
+test_case_ids = []
+for file in os.listdir(tests_path):
+    group_name = file.split(".")[0]
```
Comment (Rebits) [inline tests/integration/test_authd/force_options/test_authd_force_options.py:73]:
```suggestion
    group_name = file.split('.')[0]
```

File: tests/integration/test_authd/force_options/test_authd_force_options.py [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+    file_tests = read_yaml(os.path.join(tests_path, file))
+    tests = tests + file_tests
+    test_case_ids = test_case_ids + [f"{group_name} {test_case['name']}" for test_case in file_tests]
+
+# Variables
+log_monitor_paths = []
+
+receiver_sockets_params = [(("localhost", 1515), 'AF_INET', 'SSL_TLSv1_2')]
```
Comment (Rebits) [inline tests/integration/test_authd/force_options/test_authd_force_options.py:81]:
```suggestion
receiver_sockets_params = [(('localhost', 1515), 'AF_INET', 'SSL_TLSv1_2')]
```

File: tests/integration/test_authd/force_options/test_authd_force_options_invalid_config.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+# Configurations
+configurations = load_wazuh_configurations(configurations_path, __name__)
+local_internal_options = {'authd.debug': '2'}
+
+# Tests
+tests = []
+test_case_ids = []
+for file in os.listdir(tests_path):
+    group_name = file.split(".")[0]
```
Comment (Rebits) [inline tests/integration/test_authd/force_options/test_authd_force_options_invalid_config.py:71]:
```suggestion
    group_name = file.split('.')[0]
```

File: tests/integration/test_authd/force_options/test_authd_force_options_invalid_config.py [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+        Different test cases are contained in an external YAML file (invalid_config folder) which includes
+        different possible wrong settings.
+
+    expected_output:
+        - Invalid configuration values error.
+    '''
+
+    truncate_file(LOG_FILE_PATH)
+    try:
+        control_service("restart", daemon=DAEMON_NAME)
```
Comment (Rebits) [inline tests/integration/test_authd/force_options/test_authd_force_options_invalid_config.py:138]:
```suggestion
        control_service('restart', daemon=DAEMON_NAME)
```

**Review by Rebits — APPROVED** [review]
GJ, some changes should be implemented in order to improve readability or fit with the repository standards. Due to the urgency of this development, these changes should be implemented in a separate Issue/PR in which we could solve other `test_authd` module errors.

**Comment by DProvinciani** [comment]
All the suggestions made by the reviewers, as they are not critical, will be implemented as part of the issue https://github.com/wazuh/wazuh-qa/issues/2184 after integrating this PR.

**Comment by snaow** [comment]
LGTM. GJ Everyone.

File: deps/wazuh_testing/wazuh_testing/authd.py [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+
+
+def create_authd_request(input):
+    """
+    Creates a command to request keys to Authd.
+
+    Args:
+        input (dict): Dictionary with the content of the request command.
+    """
+    command = ""
```
[the remark this answers is not in the document]
Comment (palaciosjeremias) [reply deps/wazuh_testing/wazuh_testing/authd.py:28]: Fixed

**Pull request merged**